{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import dill\n",
    "# dill.dump_session(\"citeseq_session.db\")\n",
    "dill.load_session(\"citeseq_session.db\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib as mpl\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.cluster.hierarchy import dendrogram,linkage\n",
    "from scipy.spatial.distance import pdist,squareform"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Reprocessing CITESeq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Importing Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Source paper:\n",
    "# https://www.nature.com/articles/nmeth.4380.pdf?origin=ppub\n",
    "\n",
    "# Data is available on GEO, at GSE100866\n",
    "# https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE100866"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Loading \n",
    "\n",
    "# Numpy facilities for loading csvs are kind of dumb, so we count the number of columns in the target\n",
    "\n",
    "# columns = !head -n 1 ../../raw_data/citeseq/GSE100866_CBMC_8K_13AB_10X-RNA_umi.csv | grep -o ',' | wc -l\n",
    "# columns = int(columns[0]) \n",
    "\n",
    "# rows = !wc -l ../../raw_data/citeseq/GSE100866_CBMC_8K_13AB_10X-RNA_umi.csv \n",
    "# rows = int(rows[0].split()[0]) - 1\n",
    "\n",
    "columns = !head -n 1 ../../raw_data/citeseq/GSE100866_PBMC_vs_flow_10X-RNA_umi.csv | grep -o ',' | wc -l\n",
    "columns = int(columns[0]) \n",
    "\n",
    "rows = !wc -l ../../raw_data/citeseq/GSE100866_PBMC_vs_flow_10X-RNA_umi.csv\n",
    "rows = int(rows[0].split()[0]) - 1\n",
    "\n",
    "\n",
    "print(f\"Rows: {rows}, Columns: {columns}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "umi_counts = np.zeros((rows,columns),dtype=int)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# with open('../../raw_data/citeseq/GSE100866_CBMC_8K_13AB_10X-RNA_umi.csv') as file:\n",
    "with open('../../raw_data/citeseq/GSE100866_PBMC_vs_flow_10X-RNA_umi.csv') as file:\n",
    "    file.seek(0)\n",
    "    lines = []\n",
    "    while True:\n",
    "        try:\n",
    "            line = file.readline()\n",
    "            if len(line) < 1:\n",
    "                break\n",
    "            lines.append(line)\n",
    "#             print(len(lines))\n",
    "        except:\n",
    "            break\n",
    "\n",
    "            \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i,line in enumerate(lines[1:]):\n",
    "#     print(i)\n",
    "    split = [int(x) for x in line.split(',')[1:]]\n",
    "    umi_counts[i] = np.array(split)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_header = []\n",
    "\n",
    "for i,line in enumerate(lines[1:]):\n",
    "#     print(i)\n",
    "    gene_header.append(line.split(',')[0].strip(\"\\\"\"))\n",
    "    \n",
    "gene_header = np.array(gene_header)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "rna_cell_header = [id.strip(\"\\\"\\n\") for id in lines[0].split(',')[1:]]\n",
    "rna_cell_header"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Saving the separated files in a numpy-friendly format\n",
    "\n",
    "# np.savetxt('../../raw_data/citeseq/GSE100866_CBMC_umi_counts.tsv',umi_counts,fmt='%i')\n",
    "# np.savetxt('../../raw_data/citeseq/GSE100866_CBMC_gene_header.tsv',gene_header,fmt='%s')\n",
    "\n",
    "np.savetxt('../../raw_data/citeseq/GSE100866_PBMC_umi_counts.tsv',umi_counts,fmt='%i')\n",
    "np.savetxt('../../raw_data/citeseq/GSE100866_PBMC_gene_header.tsv',gene_header,fmt='%s')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# with open('../../raw_data/citeseq/GSE100866_CBMC_8K_13AB_10X-ADT_umi.csv') as file:\n",
    "with open('../../raw_data/citeseq/GSE100866_PBMC_vs_flow_10X-ADT_umi.csv') as file:\n",
    "    file.seek(0)\n",
    "    lines = []\n",
    "    while True:\n",
    "        try:\n",
    "            line = file.readline()\n",
    "            if len(line) < 1:\n",
    "                break\n",
    "            lines.append(line)\n",
    "#             print(len(lines))\n",
    "        except:\n",
    "            break\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "fluorescence_cell_header = [id.strip(\"\\\"\\n\") for id in lines[0].split(',')[1:]]\n",
    "fluorescence_cell_header"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fluorescence_umis = np.zeros((len(lines)-1,columns))\n",
    "\n",
    "for i,line in enumerate(lines[1:]):\n",
    "#     print(i)\n",
    "    split = [int(x) for x in line.split(',')[1:]]\n",
    "    fluorescence_umis[i] = np.array(split)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fluorescence_header = []\n",
    "\n",
    "for i,line in enumerate(lines[1:]):\n",
    "#     print(i)\n",
    "    fluorescence_header.append(line.split(',')[0].strip(\"\\\"\"))\n",
    "    \n",
    "fluorescence_header = np.array(fluorescence_header)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fluorescence_header"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# np.savetxt('../../raw_data/citeseq/GSE100866_CBMC_fluorescence_umis.tsv',fluorescence_umis,fmt='%i')\n",
    "# np.savetxt('../../raw_data/citeseq/GSE100866_CBMC_fluorescence_header.tsv',fluorescence_header,fmt='%s')\n",
    "\n",
    "np.savetxt('../../raw_data/citeseq/GSE100866_PBMC_fluorescence_umis.tsv',fluorescence_umis,fmt='%i')\n",
    "np.savetxt('../../raw_data/citeseq/GSE100866_PBMC_fluorescence_header.tsv',fluorescence_header,fmt='%s')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Here we check just in case that rna and fluorescence cell headers are identical. All good.\n",
    "\n",
    "fluorescence_cell_header == rna_cell_header"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.savetxt('../../raw_data/citeseq/GSE100866_PBMC_cell_header.tsv',rna_cell_header,fmt='%s')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Only applicable to CBMC\n",
    "\n",
    "# Finally CCR7 is not actually present (all values 0), so let's truncate \n",
    "\n",
    "# fluorescence_umis = fluorescence_umis[:13]\n",
    "# fluorescence_header = fluorescence_header[:13]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## In case we need to reload, after clearing the string objects from memory:\n",
    "\n",
    "umi_counts = np.loadtxt('../../raw_data/citeseq/GSE100866_PBMC_umi_counts.tsv')\n",
    "fluorescence_umis = np.loadtxt('../../raw_data/citeseq/GSE100866_PBMC_fluorescence_umis.tsv')\n",
    "\n",
    "gene_header = np.loadtxt('../../raw_data/citeseq/GSE100866_PBMC_gene_header.tsv',dtype=str)\n",
    "fluorescence_header = np.loadtxt('../../raw_data/citeseq/GSE100866_PBMC_fluorescence_header.tsv',dtype=str)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Paring Down Data To Relevant Elements"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# First we will need to interpret umi counts as normalized expression values. Many approaches are used here\n",
    "\n",
    "# The source paper performed extremely basic normalization without deconvolution or distribution compensation\n",
    "# This is fine for a preliminary analysis, I don't feel like importing stuff so let's keep it basic:\n",
    "\n",
    "umi_sums = np.sum(umi_counts,axis=0)\n",
    "\n",
    "log_size_normalized = np.log10(umi_counts*10000 / np.tile(umi_sums,(umi_counts.shape[0],1)))\n",
    "log_size_normalized[umi_counts == 0] = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We want to check how many genes are meaningfully expressed in the data set,\n",
    "# and how variant they are:\n",
    "\n",
    "\n",
    "means = np.mean(log_size_normalized,axis=1)\n",
    "variance = np.var(log_size_normalized,axis=1)\n",
    "\n",
    "mean_ranking = np.argsort(means)\n",
    "variance_ranking = np.argsort(variance)\n",
    "\n",
    "plt.figure()\n",
    "plt.title(\"Means Vs Variance For All Genes\")\n",
    "plt.scatter(means,variance,s=1)\n",
    "plt.xlabel(\"Means\")\n",
    "plt.ylabel(\"Variances\")\n",
    "plt.show()\n",
    "\n",
    "plt.figure()\n",
    "plt.title(\"Mean-ranked Variance\")\n",
    "plt.scatter(np.arange(5000),variance[mean_ranking[-5000:]],s=1,c=means[mean_ranking[-5000:]],cmap='binary')\n",
    "plt.colorbar()\n",
    "plt.show()\n",
    "\n",
    "plt.figure()\n",
    "plt.title(\"Sparsity\")\n",
    "plt.scatter(np.arange(20000),(np.sum(umi_counts > 0,axis=1)/umi_counts.shape[1])[mean_ranking[-20000:]],s=1)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here we notice that there are two distinct apparent populations of genes. We recall that the source paper spiked in mouse genetic material, and we would like to eliminate the mouse genes from further analysis:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### PBMC:\n",
    "#Mouse genes appear after position 17014 in the gene header.\n",
    "print(list(gene_header[:17014]))\n",
    "\n",
    "\n",
    "### CBMC:\n",
    "# Mouse genes appear after position 20401 in the gene header.\n",
    "# print(list(gene_header[:20401]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "human_normalized = log_size_normalized[:17014]\n",
    "human_gene_header = gene_header[:17014]\n",
    "\n",
    "# human_normalized = log_size_normalized[:20401]\n",
    "# human_gene_header = gene_header[:20401]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We also want to eliminate mouse cells from the analysis (we can examine them later if needed)\n",
    "# The source paper uses a criterion that more than 90% of UMIs in a cell must be aligned to human genes for a cell to be \n",
    "# considered human, so let's stick with that. \n",
    "\n",
    "umi_sums = np.sum(umi_counts,axis=0)\n",
    "umi_sums.shape\n",
    "\n",
    "umi_human_gene_sums = np.sum(umi_counts[:17014],axis=0)\n",
    "\n",
    "human_fraction = umi_human_gene_sums/umi_sums\n",
    "\n",
    "human_mask = human_fraction > .9\n",
    "\n",
    "print(np.sum(human_mask))\n",
    "\n",
    "plt.figure()\n",
    "plt.hist(umi_human_gene_sums,bins=50)\n",
    "plt.show()\n",
    "\n",
    "plt.figure()\n",
    "plt.hist(umi_sums,bins=50)\n",
    "plt.show()\n",
    "\n",
    "plt.figure()\n",
    "plt.hist(human_fraction,bins=50)\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "human_normalized = human_normalized.T[human_mask].T\n",
    "fluorescence_umis = fluorescence_umis.T[human_mask].T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Rerunning earlier analysis\n",
    "\n",
    "\n",
    "means = np.mean(human_normalized,axis=1)\n",
    "variance = np.var(human_normalized,axis=1)\n",
    "\n",
    "mean_ranking = np.argsort(means)\n",
    "variance_ranking = np.argsort(variance)\n",
    "\n",
    "plt.figure()\n",
    "plt.title(\"Means Vs Variance For All Genes\")\n",
    "plt.scatter(means,variance,s=1)\n",
    "plt.xlabel(\"Means\")\n",
    "plt.ylabel(\"Variances\")\n",
    "plt.show()\n",
    "\n",
    "plt.figure()\n",
    "plt.title(\"Mean-ranked Variance\")\n",
    "plt.scatter(np.arange(5000),variance[mean_ranking[-5000:]],s=1,c=means[mean_ranking[-5000:]],cmap='binary')\n",
    "plt.colorbar()\n",
    "plt.show()\n",
    "\n",
    "plt.figure()\n",
    "plt.title(\"Variance-ranked Variance\")\n",
    "plt.scatter(np.arange(5000),variance[variance_ranking[-5000:]],s=1,c=means[variance_ranking[-5000:]],cmap='binary')\n",
    "plt.colorbar()\n",
    "plt.show()\n",
    "\n",
    "plt.figure()\n",
    "plt.title(\"Sparsity\")\n",
    "plt.scatter(np.arange(5000),(np.sum(umi_counts[:20401] > 0,axis=1)/umi_counts[:20401].shape[1])[mean_ranking[-5000:]],s=1)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## We also want to examine the coefficient of variance to look at genes that may meaningfully contribute to differentiation \n",
    "\n",
    "cov = means[variance_ranking[-5000:]] / variance[variance_ranking[-5000:]]\n",
    "\n",
    "plt.figure()\n",
    "plt.title(\"CoV for top genes\")\n",
    "plt.scatter(np.arange(5000),cov,s=1)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In all likelihood we are safe in paring down the output to only the top 2000 most variant genes. This will probably leave out transcription factors and the like, but... vOv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "truncated_counts = human_normalized[variance_ranking[-2000:]].T\n",
    "header = human_gene_header[variance_ranking[-2000:]]\n",
    "\n",
    "# truncated_counts = human_normalized[mean_ranking[-7000:]].T\n",
    "# header = human_gene_header[mean_ranking[-7000:]]\n",
    "\n",
    "from sklearn.preprocessing import scale\n",
    "\n",
    "counts = scale(truncated_counts)\n",
    "\n",
    "print(counts.shape)\n",
    "print(header.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Let's normalize the antibody derived tags as well\n",
    "# Source paper calls for normalization via Centered Log Ratio\n",
    "\n",
    "from scipy.stats.mstats import gmean\n",
    "\n",
    "f_gmeans = gmean(fluorescence_umis+1,axis=0)\n",
    "# f_gmeans\n",
    "fluorescence_counts = np.log((fluorescence_umis + 1) / np.tile(f_gmeans,(fluorescence_umis.shape[0],1)))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.savetxt(\"citeseq_counts.tsv\",counts)\n",
    "np.savetxt(\"citeseq_adt.tsv\",fluorescence_counts)\n",
    "np.savetxt(\"citeseq_header.txt\",header,fmt='%s')\n",
    "np.savetxt(\"citeseq_fluorescence_header.tsv\",fluorescence_header,fmt='%s')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "# counts = np.loadtxt('../../raw_data/citeseq_counts.tsv')\n",
    "# fluorescence_counts = np.loadtxt('../../raw_data/citeseq_adt.tsv')\n",
    "# header = np.loadtxt('../../raw_data/citeseq_header.txt',dtype=str)\n",
    "# fluorescence_header = np.loadtxt('../../raw_data/citeseq_fluorescence_header.txt',dtype=str)\n",
    "counts = np.loadtxt('citeseq_counts.tsv')\n",
    "fluorescence_counts = np.loadtxt('citeseq_adt.tsv')\n",
    "header = np.loadtxt('citeseq_header.txt',dtype=str)\n",
    "fluorescence_header = np.loadtxt('citeseq_fluorescence_header.tsv',dtype=str)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Processing Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# First let's take a quick look at the fluorescence and genetic expression data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.manifold import TSNE\n",
    "from sklearn.decomposition import PCA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fluorescence_coordinates = TSNE().fit_transform(fluorescence_counts.T)\n",
    "\n",
    "plt.figure()\n",
    "plt.title(\"TSNE Manifold Mapping of Fluorescence Values for Blood Cells\")\n",
    "plt.scatter(fluorescence_coordinates.T[0],fluorescence_coordinates.T[1],s=1)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_coordinates = TSNE().fit_transform(PCA(n_components=20).fit_transform(counts))\n",
    "# gene_coordinates = TSNE(metric='cosine').fit_transform(counts)\n",
    "\n",
    "\n",
    "plt.figure()\n",
    "plt.title(\"TSNE Manifold Mapping of PCA Decomposed Gene Expression Values, Blood Cells \")\n",
    "plt.scatter(gene_coordinates.T[0],gene_coordinates.T[1],s=1)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Now we need to recreate the canonical labels based on antibody surface markers. We have markers as follows:\n",
    "print(fluorescence_header)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# Agglomerative clustering is also going to be a helpful way of looking at the data:\n",
    "\n",
    "from scipy.cluster.hierarchy import dendrogram,linkage\n",
    "\n",
    "feature_sort = dendrogram(linkage(counts.T,metric='cosine',method='average'),no_plot=True)['leaves']\n",
    "sample_sort = dendrogram(linkage(counts,metric='cosine',method='average'),no_plot=True)['leaves']\n",
    "fluorescence_sort = dendrogram(linkage(fluorescence_counts,method='average'),no_plot=True)['leaves']\n",
    "\n",
    "plt.figure(figsize=(20,20))\n",
    "plt.imshow(counts[sample_sort].T[feature_sort].T,cmap='bwr',aspect='auto')\n",
    "plt.show()\n",
    "\n",
    "plt.figure(figsize=(20,20))\n",
    "plt.imshow(np.log10(1 + fluorescence_counts.T[sample_sort]),aspect='auto')\n",
    "plt.colorbar()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# We can map cell surface markers to cell identities using refernce literature like the BD cell surface marker handbook:\n",
    "# https://www.bdbiosciences.com/documents/cd_marker_handbook.pdf\n",
    "\n",
    "# A basic mapping would be:\n",
    "\n",
    "# CD3   + ::: T Cells\n",
    "# CD2   + ::: T/B/NK (Adhesion)\n",
    "# CD4   + ::: Granular cells (T, Macro, Granulo, early phase)\n",
    "\n",
    "# CD8   + ::: T/NK (unclear since handbook lists 2 variants)\n",
    "# CD45  + ::: T/B cell differentiation (not erythrocyte)\n",
    "\n",
    "# CD56  + ::: T and NK (adhesion)\n",
    "# CD57  + ::: T and NK (adhesion)\n",
    "# CD16  + ::: T, Dendritic, NK, Macro, Granulo (IgG binding)\n",
    "# CD10  + ::: NK, Endothelial \n",
    "\n",
    "# CD11c + ::: General marker (inflammatory response?)\n",
    "# CD14  + ::: LPS response, Macrophage, Granulocyte\n",
    "\n",
    "\n",
    "# CD19  + ::: B Cell, Dendritic, Stem marker\n",
    "# CD34  + ::: Stem Cells, Endothelial \n",
    "# CCR5  + ::: Migration (T, Macro, Gran)\n",
    "# CCR7  + ::: T/B/Dend\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We would like to cluster the cells, then examine the fluorescence value of each marker in each cluster\n",
    "\n",
    "print(fluorescence_header)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure()\n",
    "plt.title(\"All Cells\")\n",
    "plt.scatter(fluorescence_counts[9],fluorescence_counts[0],s=1)\n",
    "plt.xlabel(\"CD19 (TPM)\")\n",
    "plt.ylabel(\"CD3 (TPM)\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cd3_p_mask = fluorescence_counts[0] > 1.2\n",
    "cd19_p_mask = fluorescence_counts[9] > 1\n",
    "\n",
    "t_mask = np.logical_and(cd3_p_mask,np.logical_not(cd19_p_mask))\n",
    "b_mask = np.logical_and(cd19_p_mask,np.logical_not(cd3_p_mask))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure()\n",
    "plt.title(\"All Cells\")\n",
    "plt.scatter(fluorescence_counts[9],fluorescence_counts[0],s=1,c=b_mask)\n",
    "plt.xlabel(\"CD19\")\n",
    "plt.ylabel(\"CD3\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now we'd like to differentiate between CD4 and CD8 T cells:\n",
    "\n",
    "plt.figure()\n",
    "plt.title(\"T Cells only (CD3+,CD19-)\")\n",
    "plt.scatter(fluorescence_counts[2][t_mask],fluorescence_counts[1][t_mask],s=1)\n",
    "plt.xlabel(\"CD8A\")\n",
    "plt.ylabel(\"CD4\")\n",
    "plt.show()\n",
    "\n",
    "cd4_t = np.logical_and(t_mask,np.logical_and(fluorescence_counts[1] > 0,fluorescence_counts[2] < 1.5))\n",
    "cd8_t = np.logical_and(t_mask,np.logical_and(fluorescence_counts[1] < 0,fluorescence_counts[2] > 3))\n",
    "\n",
    "plt.figure()\n",
    "plt.title(\"T Cells only (CD3+,CD19-)\")\n",
    "plt.scatter(fluorescence_counts[2][t_mask],fluorescence_counts[1][t_mask],c=cd4_t[t_mask],s=1)\n",
    "plt.xlabel(\"CD8A\")\n",
    "plt.ylabel(\"CD4\")\n",
    "plt.show()\n",
    "\n",
    "plt.figure()\n",
    "plt.title(\"T Cells only (CD3+,CD19-)\")\n",
    "plt.scatter(fluorescence_counts[2][t_mask],fluorescence_counts[1][t_mask],c=cd8_t[t_mask],s=1)\n",
    "plt.xlabel(\"CD8A\")\n",
    "plt.ylabel(\"CD4\")\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Now let's examine how these are distributed in the TSNE:\n",
    "\n",
    "plt.figure()\n",
    "plt.title(\"TSNE Manifold Mapping of PCA Decomposed Gene Expression Values, CD4+ T Cells Colored, Blood Cells \")\n",
    "plt.scatter(gene_coordinates.T[0],gene_coordinates.T[1],c=cd4_t,s=1)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure()\n",
    "plt.title(\"TSNE Manifold Mapping of PCA Decomposed Gene Expression Values, CD8+ T Cells Colored, Blood Cells \")\n",
    "plt.scatter(gene_coordinates.T[0],gene_coordinates.T[1],c=cd8_t,s=1)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Finally let's try to map out naive, memory, and effector cells\n",
    "\n",
    "plt.figure()\n",
    "plt.title(\"CD4 T Cells\")\n",
    "plt.scatter(fluorescence_counts[3][cd4_t],fluorescence_counts[5][cd4_t],s=1)\n",
    "plt.xlabel(\"CD2\")\n",
    "plt.ylabel(\"CD57\")\n",
    "plt.show()\n",
    "\n",
    "plt.figure()\n",
    "plt.title(\"CD8 T Cells\")\n",
    "plt.scatter(fluorescence_counts[3][cd8_t],fluorescence_counts[5][cd8_t],s=1)\n",
    "plt.xlabel(\"CD2\")\n",
    "plt.ylabel(\"CD57\")\n",
    "plt.show()\n",
    "\n",
    "cd4_t_naive = np.logical_and(cd4_t,np.logical_and(fluorescence_counts[3] < 2.5,fluorescence_counts[5] < 1))\n",
    "cd4_t_57_effector = np.logical_and(cd4_t,fluorescence_counts[5] > 1)\n",
    "cd4_t_memory = np.logical_and(cd4_t,np.logical_and(fluorescence_counts[3] > 2.5,fluorescence_counts[5] < 1))\n",
    "\n",
    "cd8_t_naive = np.logical_and(cd8_t,np.logical_and(fluorescence_counts[3] < 2.5,fluorescence_counts[5] < 1))\n",
    "cd8_t_57_effector = np.logical_and(cd8_t,fluorescence_counts[5] > 1)\n",
    "cd8_t_memory = np.logical_and(cd8_t,np.logical_and(fluorescence_counts[3] > 2.5,fluorescence_counts[5] < 1))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure()\n",
    "plt.title(\"CD4 T Cells\")\n",
    "plt.scatter(fluorescence_counts[3][cd4_t],fluorescence_counts[5][cd4_t],c=cd4_t_57_effector[cd4_t],s=1)\n",
    "plt.xlabel(\"CD2\")\n",
    "plt.ylabel(\"CD57\")\n",
    "plt.show()\n",
    "\n",
    "plt.figure()\n",
    "plt.title(\"CD8 T Cells\")\n",
    "plt.scatter(fluorescence_counts[3][cd8_t],fluorescence_counts[5][cd8_t],c=cd8_t_57_effector[cd8_t],s=1)\n",
    "plt.xlabel(\"CD2\")\n",
    "plt.ylabel(\"CD57\")\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fc = forest.coordinates()\n",
    "\n",
    "plt.figure()\n",
    "plt.scatter(fc[:,0],fc[:,1],s=1,c=cd4_t_naive)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Forest Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append(\"../src/\")\n",
    "import lumberjack\n",
    "import tree_reader_sc as tr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# forest = lumberjack.fit(input_counts=counts,output_counts=counts,header=header,ifs=400,ofs=400,ss=400,trees=50,depth=5,norm='l1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# forest.backup('citeseq_forest_l1')\n",
    "# forest.backup('citeseq_forest_l1_cache')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# !ls -lh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "forest = tr.Forest.reconstitute('citeseq_forest_l1_cache')\n",
    "# forest = tr.Forest.reconstitute('citeseq_forest')\n",
    "# forest = tr.Forest.reconstitute('citeseq_forest_cache')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Removing:['sample_cluster_0' 'sample_cluster_1' 'sample_cluster_2'\n",
      " 'sample_cluster_3' 'sample_cluster_4' 'sample_cluster_5'\n",
      " 'sample_cluster_6' 'sample_cluster_7' 'sample_cluster_8'\n",
      " 'sample_cluster_9' 'sample_cluster_10' 'sample_cluster_11'\n",
      " 'sample_cluster_12' 'sample_cluster_13' 'sample_cluster_14'\n",
      " 'sample_cluster_15' 'sample_cluster_16' 'sample_cluster_17'\n",
      " 'sample_cluster_18' 'sample_cluster_19' 'sample_cluster_20'\n",
      " 'sample_cluster_21' 'sample_cluster_22' 'sample_cluster_23']\n"
     ]
    }
   ],
   "source": [
    "forest.reset_output_featuers()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# forest.set_cache(True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "## First we must find common marginal effects of splits in the random forest.\n",
    "# We can do this by clustering vectors representing the marginal effect of each split\n",
    "\n",
    "# forest.reset_clusters()\n",
    "# forest.interpret_splits(depth=5,mode='additive',sub=.5,metric='cosine',k=10,pca=False,override=False)\n",
    "\n",
    "## TO DO: REDO DEFAULTS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'Forest' object has no attribute 'split_cluster_transitions'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-9-4e74f826443a>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;31m# print(forest.split_clusters[0].css_additive_panel(features=forest.output_features[1990:2000]))\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;31m# forest.split_clusters[1].json_cluster_summary(n=5)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 5\u001b[0;31m \u001b[0mforest\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhtml_tree_summary\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/taylor/vision/rust_prototype/rusty_lumberjack/src/tree_reader_sc.py\u001b[0m in \u001b[0;36mhtml_tree_summary\u001b[0;34m(self, n, type, custom, labels, features, primary, cmap, secondary, figsize)\u001b[0m\n\u001b[1;32m   2831\u001b[0m                     \u001b[0;31m# matrix\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2832\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2833\u001b[0;31m                     \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit_cluster_transitions\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mj\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2834\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2835\u001b[0m                         \u001b[0;31m# If the transitions are non-zero we obtain the coordinates\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'Forest' object has no attribute 'split_cluster_transitions'"
     ]
    }
   ],
   "source": [
    "## Placeholder debugging cell\n",
    "# forest.plot_tree_summary(figsize=(30,10),type='additive_features',features=forest.output_features[1990:2000],cmap='bwr')\n",
    "# print(forest.split_clusters[0].css_additive_panel(features=forest.output_features[1990:2000]))\n",
    "# forest.split_clusters[1].json_cluster_summary(n=5)\n",
    "forest.html_tree_summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "forest.dependence_tree()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "## We want to examine the common splits in two ways: \n",
    "\n",
    "# Cell populations they affect and effects of each split\n",
    "\n",
    "forest.plot_tree_summary(figsize=(20,6),type='score')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "\n",
    "tc = forest.coordinates(no_plot=True)\n",
    "\n",
    "for split_cluster in forest.split_clusters:\n",
    "    print(split_cluster.id)\n",
    "    try:\n",
    "        print(sorted(list(split_cluster.braid_features().items()),key=lambda x:x[1])[::-1])\n",
    "#         print(list(split_cluster.braid_features().items()))\n",
    "    except:\n",
    "        pass\n",
    "#     split_cluster.plot_cell_counts()\n",
    "#     cell_score = split_cluster.cell_counts()\n",
    "\n",
    "#     f = plt.figure(figsize=(15,10))\n",
    "#     plt.title(\"Cell Score\")\n",
    "#     plt.scatter(tc[:,0],tc[:,1],c=cell_score)\n",
    "#     plt.colorbar()\n",
    "#     plt.show()\n",
    "    \n",
    "    sister_color = split_cluster.sister_scores()\n",
    "\n",
    "    f = plt.figure(figsize=(15,10))\n",
    "    plt.title(\"Sister Split\")\n",
    "    plt.scatter(tc[:,0],tc[:,1],c=sister_color,cmap='bwr',vmin=max(sister_color)*-1)\n",
    "    plt.colorbar()\n",
    "#     plt.show()\n",
    "    plt.savefig(f\"ss{split_cluster.id}.png\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# forest.cluster_samples_simple(sub=.5,k=20,pca=20,metric='cosine',override=True)\n",
    "forest.cluster_samples_encoding(sub=.3,metric='cosine',k=20,override=True,pca=20)\n",
    "len(forest.sample_clusters)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "forest.plot_cell_clusters()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure()\n",
    "plt.title(\"TSNE Manifold Mapping of Fluorescence Values for Blood Cells\")\n",
    "plt.scatter(fluorescence_coordinates.T[0],fluorescence_coordinates.T[1],s=1,c=forest.sample_labels,cmap='rainbow')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from scipy.cluster.hierarchy import dendrogram,linkage\n",
    "\n",
    "feature_sort = dendrogram(linkage(counts.T,metric='cosine',method='average'),no_plot=True)['leaves']\n",
    "sample_sort = np.argsort(forest.sample_labels)\n",
    "\n",
    "plt.figure(figsize=(20,20))\n",
    "plt.imshow(counts[sample_sort].T[feature_sort].T,cmap='bwr',aspect='auto')\n",
    "plt.show()\n",
    "\n",
    "plt.figure(figsize=(20,20))\n",
    "plt.imshow(fluorescence_counts.T[sample_sort],aspect='auto')\n",
    "plt.xticks(np.arange(10),fluorescence_header)\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fluorescence_counts.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_cluster_features = forest.sample_cluster_feature_matrix()\n",
    "sample_cluster_sort = dendrogram(linkage(sample_cluster_features,metric='cosine',method='average'),no_plot=True)['leaves']\n",
    "\n",
    "\n",
    "plt.figure(figsize=(20,20))\n",
    "plt.imshow(sample_cluster_features[sample_cluster_sort].T[feature_sort].T,cmap='bwr',aspect='auto')\n",
    "plt.yticks(np.arange(len(forest.sample_clusters)),np.arange(len(forest.sample_clusters))[sample_cluster_sort])\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# focused = sample_cluster_features[sample_cluster_sort].T[feature_sort].T[-4:]\n",
    "# feature_resort = dendrogram(linkage(focused.T,metric='cosine',method='average'),no_plot=True)['leaves']\n",
    "\n",
    "# plt.figure(figsize=(20,5))\n",
    "# plt.imshow(focused,cmap='bwr',aspect='auto')\n",
    "# plt.show()\n",
    "\n",
    "# plt.figure(figsize=(5,30))\n",
    "# plt.imshow(focused.T[feature_resort],cmap='bwr',aspect='auto')\n",
    "# plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "t_markers = np.array([\n",
    "    b_mask,\n",
    "    t_mask,\n",
    "    cd4_t,\n",
    "    cd8_t,\n",
    "    cd4_t_naive,\n",
    "    cd4_t_memory,\n",
    "    cd4_t_57_effector,\n",
    "    cd8_t_naive,\n",
    "    cd8_t_memory,\n",
    "    cd8_t_57_effector])\n",
    "\n",
    "t_marker_header = [\n",
    "    \"b_cell\",\n",
    "    \"t_cell\",\n",
    "    \"cd4_t_cell\",\n",
    "    \"cd8_t_cell\",\n",
    "    \"cd4_t_naive\",\n",
    "    \"cd4_t_memory\",\n",
    "    \"cd4_t_effector\",\n",
    "    \"cd8_t_naive\",\n",
    "    \"cd8_t_memory\",\n",
    "    \"cd8_t_effector\",\n",
    "]\n",
    "\n",
    "# t_markers.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(20,20))\n",
    "plt.imshow(t_markers.T[sample_sort],aspect='auto')\n",
    "plt.xticks(np.arange(10),fluorescence_header)\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## We would now like to see what kind of distribution there is between forest splits and sample clusters\n",
    "# for the fluorescence features\n",
    "\n",
    "fluorescence_features = list(fluorescence_header) + list(t_marker_header)\n",
    "fluorescence_feature_values = np.array([f for f in fluorescence_counts[:-1]] + [m for m in t_markers])\n",
    "\n",
    "print(fluorescence_features)\n",
    "print(len(fluorescence_features))\n",
    "print(fluorescence_feature_values.shape)\n",
    "\n",
    "for feature,feature_values in zip(fluorescence_features,fluorescence_feature_values):\n",
    "    forest.add_output_feature(feature_values,feature)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "forest.output_features[2000:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "sample_cluster_fluorescence = forest.sample_cluster_feature_matrix(fluorescence_features)\n",
    "# print(sample_cluster_fluorescence.shape)\n",
    "\n",
    "plt.figure(figsize=(20,20))\n",
    "plt.imshow(sample_cluster_fluorescence[:,10:],aspect='auto')\n",
    "plt.yticks(np.arange(len(forest.sample_clusters)),np.arange(len(forest.sample_clusters)))\n",
    "plt.xticks(np.arange(10),fluorescence_features[10:])\n",
    "plt.colorbar()\n",
    "plt.show()\n",
    "\n",
    "# plt.figure(figsize=(20,20))\n",
    "# plt.imshow(sample_cluster_fluorescence[sample_cluster_sort],aspect='auto')\n",
    "# plt.yticks(np.arange(23),np.arange(23)[sample_cluster_sort])\n",
    "# plt.xticks(np.arange(20),fluorescence_features)\n",
    "# plt.colorbar()\n",
    "# plt.show()\n",
    "\n",
    "# plt.figure(figsize=(20,20))\n",
    "# plt.imshow(sample_cluster_fluorescence[sample_cluster_sort][:,:10],aspect='auto')\n",
    "# plt.yticks(np.arange(23),np.arange(23)[sample_cluster_sort])\n",
    "# plt.xticks(np.arange(10),fluorescence_features[:10])\n",
    "# plt.colorbar()\n",
    "# plt.show()\n",
    "\n",
    "plt.figure(figsize=(20,20))\n",
    "plt.imshow(sample_cluster_fluorescence[sample_cluster_sort][:,10:],aspect='auto')\n",
    "plt.yticks(np.arange(len(forest.sample_clusters)),np.arange(len(forest.sample_clusters))[sample_cluster_sort])\n",
    "plt.xticks(np.arange(10),fluorescence_features[10:])\n",
    "plt.colorbar()\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "forest.likely_tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def flatten_tree(tree):\n",
    "    flat = []\n",
    "    for child in tree[1]:\n",
    "        flat.extend(flatten_tree(child))\n",
    "    flat.append([tree[0],[c[0] for c in tree[1]]])\n",
    "    return flat\n",
    "\n",
    "def levels(tree,level=0):\n",
    "    l = []\n",
    "    for child in tree[1]:\n",
    "        l.extend(levels(child,level=level+1))\n",
    "    l.append(level)\n",
    "    return l\n",
    "\n",
    "flat = flatten_tree(forest.likely_tree)\n",
    "\n",
    "flat_label = [f[0] for f in flat]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(flat)\n",
    "print(flat_label)\n",
    "print(np.arange(25)[flat_label])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "split_cluster_fluorescence = forest.split_cluster_feature_matrix(fluorescence_features)\n",
    "split_cluster_fluorescence.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plt.figure(figsize=(20,20))\n",
    "# plt.imshow(split_cluster_fluorescence[:,10:],aspect='auto')\n",
    "# plt.yticks(np.arange(21),np.arange(21))\n",
    "# plt.xticks(np.arange(10),fluorescence_features[:10])\n",
    "# plt.colorbar()\n",
    "# plt.show()\n",
    "\n",
    "plt.figure(figsize=(20,20))\n",
    "plt.imshow(split_cluster_fluorescence[flat_label][:,10:],aspect='auto')\n",
    "plt.yticks(np.arange(21),flat_label)\n",
    "plt.xticks(np.arange(10),fluorescence_features[10:])\n",
    "plt.colorbar()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def flat_linkage(tree):\n",
    "    flat_tree = flatten_tree(tree)\n",
    "    print(flat_tree)\n",
    "    node_translation = {n[0]:i for i,n in enumerate(flat_tree)}\n",
    "    print(node_translation)\n",
    "    node_levels = levels(tree)\n",
    "    \n",
    "    coordinates = np.zeros((len(flat_tree),2))\n",
    "    \n",
    "    coordinates[:,1] = np.arange(len(flat_tree),0,-1)\n",
    "    coordinates[:,0] = node_levels\n",
    "    \n",
    "    coordinates[:,1] /= np.max(coordinates[:,1])\n",
    "    coordinates[:,0] /= np.max(coordinates[:,0])\n",
    "    \n",
    "    links = []\n",
    "    \n",
    "    for node in flat_tree:\n",
    "        node_index = node_translation[node[0]]\n",
    "        for child in (node[1]):\n",
    "            child_index = node_translation[child]\n",
    "            links.append(([node[0],child],[coordinates[node_index,0],coordinates[child_index,0]],[coordinates[node_index,1],coordinates[child_index,1]]))\n",
    "    return links"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "flat_linkage(forest.likely_tree)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(20,20))\n",
    "ax1 = fig.add_subplot(121)\n",
    "links = flat_linkage(forest.likely_tree)\n",
    "for i,link in enumerate(links):\n",
    "#     ax1.scatter(link[2][0],link[2][1])\n",
    "    print(link[1][0],link[1][1],link[0][0],link[0][1])\n",
    "    ax1.text(link[1][1] + .1 ,link[2][1],s=link[0][1],verticalalignment='center',horizontalalignment='center')\n",
    "for i,link in enumerate(links):\n",
    "    ax1.plot(link[1],link[2])\n",
    "ax1.set_axis_off()\n",
    "ax2 = fig.add_subplot(122)\n",
    "ax2.imshow(split_cluster_fluorescence[flat_label][:,10:],aspect='auto')\n",
    "ax2.set_yticks(np.arange(21))\n",
    "ax2.set_yticklabels(flat_label)\n",
    "ax2.set_xticks(np.arange(10))\n",
    "ax2.set_xticklabels(fluorescence_features[10:],rotation=45,verticalalignment='top',horizontalalignment='right')\n",
    "fig.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "ff = [fluorescence_features[10],] + fluorescence_features[14:]\n",
    "# ff\n",
    "forest.plot_tree_summary(figsize=(20,5),type='features',features=ff,cmap='bwr')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "forest.plot_tree_summary(type='id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "forest.plot_tree_summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "forest.plot_manifold()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "forest.plot_split_clusters()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "forest.coordinates(type='tsne',override=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(forest.sample_clusters))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "forest.likely_tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd = forest.partial_dependence()\n",
    "# direction = forest.directional_matrix()\n",
    "\n",
    "plt.figure(figsize=(20,20))\n",
    "plt.imshow(pd)\n",
    "plt.show()\n",
    "\n",
    "# plt.figure(figsize=(20,20))\n",
    "# plt.imshow(direction)\n",
    "# plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.argmin(pd,axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "forest.likely_tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "forest.added_features = 43"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "forest.reset_output_featuers()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "forest.output.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "forest.added_features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "forest.output.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "forest.reset_output_featuers()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "forest.reset_output_featuers()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure()\n",
    "ax = fig.add_axes([0,0,1,1])\n",
    "forest.split_clusters[4].feature_panel(ax,fluorescence_features)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "forest.reset_cache()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def recursive_nested_css(tree,i=1):\n",
    "    node_string = \"\"\"\\\n",
    "    <div class=\"node-style\",style=\"grid-column-begin:{};\">\n",
    "            <h1>{}</h1>\n",
    "            {}\n",
    "    </div>\n",
    "    \"\"\".format(i,tree[0],\"\".join([recursive_nested_css(child,i+1) for i,child in enumerate(tree[1])]))\n",
    "    \n",
    "    return node_string"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "toy_tree = [0,[[1,[]],[2,[]]]]\n",
    "print(recursive_nested_css(toy_tree))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def generate_node_feature_html(features,values,normalization=None,cmap=None):\n",
    "    \n",
    "    if cmap is None:\n",
    "        from matplotlib.cm import get_cmap\n",
    "        cmap = get_cmap('viridis')\n",
    "    \n",
    "    html_elements = [\n",
    "        \"<table>\",\n",
    "        \"<style>\",\"th,td {padding:5px;border-bottom:1px solid #ddd;}\",\"</style>\",\n",
    "        \"<tr>\",\n",
    "        \"<th>\",\"Features\",\"</th>\",\n",
    "        \"<th>\",\"Values\",\"</th>\",\n",
    "        \"</tr>\",\n",
    "    ]\n",
    "    for feature,value in zip(features,values):\n",
    "        value_color_tag = \"\"\n",
    "        if normalization is not None:\n",
    "            normed_value = normalization(value)\n",
    "            r,g,b,a = cmap(normed_value)\n",
    "            r,g,b,a = r*100,g*100,b*100,a*100\n",
    "            value_color_tag = f'style=\"background-color:rgba({r}%,{g}%,{b}%,50%);\"'\n",
    "        feature_elements = f\"\"\"\n",
    "            <tr>\n",
    "                <td>{feature}</td>\n",
    "                <td {value_color_tag}>{value}</td>\n",
    "            </td>\n",
    "        \"\"\"\n",
    "        html_elements.append(feature_elements)\n",
    "    \n",
    "    html_elements.append(\"</table>\")\n",
    "    return \"\".join(html_elements)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "features = [\"Hsp90\",\"Klu\"]\n",
    "values = [1,.5]\n",
    "vr = mpl.cm.get_cmap('viridis')\n",
    "norm = mpl.colors.DivergingNorm(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(generate_node_feature_html(features,values,cmap=vr,normalization=norm))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
